lib/tls/session.toit

```diff
@@ -534,6 +573,317 @@ class Session:
+      throw "RESUME_FAILED"
+    next_server_packet := session_.extract_first_message_
+    if next_server_packet[0] == HANDSHAKE_ and next_server_packet[5] == NEW_SESSION_TICKET_:
+      session_ticket_ = next_server_packet[9..]
+      assert:
+        message := HandshakeHeader_ next_server_packet
+        message.length == session_ticket_.size
+      handshake_hasher.add next_server_packet[5..]
+      next_server_packet = session_.extract_first_message_
+
+    // Generate new keys in a shortened handshake.
+    key_size := cipher_suite_.key_size
+    iv_size := cipher_suite_.iv_size
+    bytes_needed := 2 * (key_size + iv_size)
+    // RFC 5246 section 6.3.
+    key_data := pseudo_random_function_ bytes_needed
+        --block_size=cipher_suite_.hmac_block_size
+        --secret=master_secret_
+        --label="key expansion"
+        --seed=(server_hello.random + client_random_)
+        cipher_suite_.hmac_hasher
+    partition := partition_byte_array_ key_data [key_size, key_size, iv_size, iv_size]
+    write_key := KeyData_ --key=partition[0] --iv=partition[2] --algorithm=cipher_suite_.algorithm
+    read_key := KeyData_ --key=partition[1] --iv=partition[3] --algorithm=cipher_suite_.algorithm
+    sent = session_.writer_.write CHANGE_CIPHER_SPEC_TEMPLATE_
+    assert: sent == CHANGE_CIPHER_SPEC_TEMPLATE_.size
+    if next_server_packet.size != 6 or next_server_packet[0] != CHANGE_CIPHER_SPEC_ or next_server_packet[5] != 1:
+      throw "Peer did not accept change cipher spec"
+    server_handshake_hash := handshake_hasher.clone.get
+    // https://www.rfc-editor.org/rfc/rfc5246#section-7.4.9
+    server_finished_expected := pseudo_random_function_ 12
+        --block_size=cipher_suite_.hmac_block_size
+        --secret=master_secret_
+        --label="server finished"
+        --seed=server_handshake_hash
+        cipher_suite_.hmac_hasher
+    server_finished_expected = #[FINISHED_, 0x00, 0x00, server_finished_expected.size] + server_finished_expected
+    // The client message hash includes the server handshake message.
+    // We know what that message is going to be, so we can add it before
+    // we receive it.
+    handshake_hasher.add server_finished_expected
+    client_handshake_hash := handshake_hasher.get
+    // The client finished messages includes the hash of the server's.
+    client_finished := pseudo_random_function_ 12
+        --block_size=cipher_suite_.hmac_block_size
+        --secret=master_secret_
+        --label="client finished"
+        --seed=client_handshake_hash
+        cipher_suite_.hmac_hasher
+    client_finished = #[FINISHED_, 0x00, 0x00, client_finished.size] + client_finished
+    set_up_session_ write_key read_key client_finished server_finished_expected
+
+  set_up_session_ write_key/KeyData_ read_key/KeyData_ client_finished/ByteArray server_finished_expected/ByteArray -> none:
+    // Add message header - 1 byte for type, 3 bytes for length.
+    session_.reads_encrypted_ = true
```

Comment:
Do we need to change stuff in the session before we know we'll complete this without errors? It also feels a bit weird that the `ToitHandshake_` updates state in the session.

Maybe the handshake could just build the symmetric session and the session state and return them as a handshake result of sorts and then the session could apply it by updating its state?